src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,554 @@
+
+	comp_info(dev, "rtnr_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = rtnr_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = rtnr_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	}
+
+	return ret;
+}
+
+static int rtnr_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "rtnr_trigger() cmd: %d", cmd);
+
+	return comp_set_state(dev, cmd);
+}
+
+/* copy and process stream data from source to sink buffers */
+static int rtnr_copy(struct comp_dev *dev)
+{
+	struct list_item *blist;
+	struct comp_buffer *sink;
+	struct comp_buffer *source;
+	int frames;
+	int num_blocks;
+	int sink_bytes;
+	int source_bytes;
+	const struct audio_stream *sources_stream[RTNR_MAX_SOURCES] = { NULL };
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sources[RTNR_MAX_SOURCES] = { NULL };
+	uint32_t flags = 0;
+	int num_sources = 0;
+
+	comp_dbg(dev, "rtnr_copy()");
+
+	/* align source streams with their respective configurations */
+	list_for_item(blist, &dev->bsource_list) {
+		source = container_of(blist, struct comp_buffer, sink_list);
+		buffer_lock(source, &flags);
+		if (source->source->state == dev->state) {
+			if (num_sources < RTNR_MAX_SOURCES) {
+				sources[num_sources] = source;
+				sources_stream[num_sources] = &source->stream;
+				num_sources++;
+			} else {
+				buffer_unlock(source, flags);
+			}
+		} else {
+			buffer_unlock(source, flags);
+		}
+	}
+
+	/* check if there are any sources active */
+	if (num_sources == 0 || num_sources > 2)
+		return -EINVAL;
+
+	#if RTNR_ENABLE_PROCESSING
+	/* put dummy data into output queue*/
+	int nch = sources_stream[0]->channels;
+	int copy_size = BLOCKLENGTH*nch*sizeof(int32_t);
+
+	RTKMA_API_First_Copy((void *)&cd->rtk_agl, copy_size);
+	#endif // RTNR_ENABLE_PROCESSING
+
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	buffer_lock(sink, &flags);
+	frames = audio_stream_avail_frames(sources_stream[0], &sink->stream);
+	buffer_unlock(sources[0], flags);
+
+	/* Process integer multiple of RTNR internal block length */
+	num_blocks = frames / RTNR_BLK_LENGTH;
+	frames = num_blocks * RTNR_BLK_LENGTH;
+
+	if (cd->count) {
+		cd->count--;
+	} else {
+		for (int i = 0; i < num_sources; ++i)
+			comp_info(dev, "rtnr_copy() sources[%d]->id: %d, frames = %d",
+				i, sources[i]->id, frames);
+		cd->count = 249; // about once per second when 16kHz
+	}
+
+	if (frames) {
+		/* Run processing function */
+		cd->rtnr_func(dev, sources_stream, &sink->stream, frames);
+
+		/* Track consume and produce */
+		source_bytes = frames * audio_stream_frame_bytes(sources_stream[0]);
+		sink_bytes = frames * audio_stream_frame_bytes(&sink->stream);
+		comp_update_buffer_consume(sources[0], source_bytes);
+		comp_update_buffer_produce(sink, sink_bytes);
+	}
+
+	#if RTNR_ENABLE_PROCESSING
+	/*real process function of rtk ma*/
+	copy_size = BLOCKLENGTH*MicNum*sizeof(int32_t);
+	RTKMA_API_Process((void *)&cd->rtk_agl, copy_size, 0);
```

Comment:
what is this doing? Are the samples not already processed? If not, we should not be updating the buffers until after this.